pkg/agent/logrouter/logrouter.go

```diff
@@ -0,0 +1,199 @@
+//     a on initialization. This allows the command line flags to work)
+//  2. debug.json: This is meant as a comprehensive log, suitable for sending to support
+//  3. memory ring: This is meant to hold all recent logs. Because the debug level can be noisy
+//  4. persisted ring: This handles recent info+ logs. It is persisted to the database, and can be queried
+
+package logrouter
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	agenttypes "github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/log/locallogger"
+	"github.com/kolide/launcher/pkg/log/ringlogger"
+	"github.com/kolide/launcher/pkg/log/teelogger"
+	"github.com/kolide/launcher/pkg/persistentring"
+)
+
+type teeloggerInt interface {
+	Add(loggers ...log.Logger)
+	Log(keyvals ...interface{}) error
+}
+
+type storeInt interface {
+	agenttypes.GetterSetterDeleterIterator
+}
+
+type ringLoggerInt interface {
+	Log(keyvals ...interface{}) error
+	GetAll() ([]map[string]any, error)
+}
+
+type logRouter struct {
+	loggerTee   teeloggerInt
+	systemTee   teeloggerInt
+	logRing     ringLoggerInt // Holds the super set of logs, in memory
+	persistRing ringLoggerInt // Only info and higher, persisted to a store
+}
+
+func New(systemLogger log.Logger) (*logRouter, error) {
+	// We always have exactly one in memory ring logger. This recieves _all_ the logs. It's used to replay into new
+	// desinatations, and is expected to be a query target for recent logs
+	logRing, err := ringlogger.New(persistentring.NewInMemory(1000))
+	if err != nil {
+		return nil, fmt.Errorf("creating memory ringlogger: %w", err)
+	}
+
+	// Setup the internal logger, this rolls up all the internal destinations
+	loggerTee := teelogger.New(logRing)
+
+	// the systemLogger should copy anything it generates into the regular logger.
+	systemTee := teelogger.New(systemLogger, loggerTee)
+
+	return &logRouter{
+		loggerTee: loggerTee,
+		systemTee: systemTee,
+		logRing:   logRing,
+	}, nil
+}
+
+func (lr *logRouter) ReplaceSystemLogger(newSystemLogger log.Logger) {
+	newSystemTee := teelogger.New(newSystemLogger, lr.loggerTee)
+	lr.systemTee = newSystemTee
+}
+
+func (lr *logRouter) Log(keyvals ...interface{}) error {
+	return lr.loggerTee.Log(keyvals...)
+}
+
+func (lr *logRouter) Logger() log.Logger {
+	return lr.loggerTee
+}
+
+func (lr *logRouter) SystemLogger() log.Logger {
+	return lr.systemTee
+}
+
+func (lr *logRouter) GetRecentLogs() ([]map[string]any, error) {
+	if lr.persistRing == nil {
+		return nil, nil
+	}
+
+	return lr.persistRing.GetAll()
+}
+
+func (lr *logRouter) GetRecentDebugLogs() ([]map[string]any, error) {
+	return lr.logRing.GetAll()
+}
+
+// Replay will replay the logs from the memory ring to a new logger. Because we're serliazing everything down to a
+// string, handling the leveling is cumbersome.
+//
+// TODO: This should probably move to the underlying ringlogger
+func (lr *logRouter) Replay(logger log.Logger) error {
```

Comment:
this is really cool and will work nicely with log shipping

Reply:
I built it for that! Well, and the idea of an action triggering a log replay

Reply:
I think there are some ways we can figure out how to clean it up. What methods live where. But the bones are shaping up